File: solutions/scala/binary-search/1/src/main/scala/BinarySearch.scala
import scala.annotation.tailrec

object BinarySearch {
  def find(items: List[Int], item: Int): Option[Int] =
    @tailrec
    def search(left: Int, right: Int): Option[Int] =
      if left > right then None
      else
        val middle = left + (right - left) / 2

        items(middle) match
          case value if value == item => Some(middle)
          case value if value < item => search(middle + 1, right)
          case _ => search(left, middle - 1)

    search(0, items.length - 1)
}

File: solutions/python/binary-search/2/binary_search.py
"""Binary search implementation."""


def find(search_list, value):
    """Return the index of value in a sorted list."""
    left = 0
    right = len(search_list) - 1

    while left <= right:
        middle = (left + right) // 2
        middle_value = search_list[middle]

        if middle_value == value:
            return middle
        if value < middle_value:
            right = middle - 1
        else:
            left = middle + 1

    raise ValueError("value not in array")
